lib/facter/framework/core/fact_loaders/fact_loader.rb

```diff
@@ -19,6 +19,10 @@ def initialize
     end
 
     def load(user_query, options)
+      if ENV['INSIDE_FACTER']
+        @log.debug('INSIDE_FACTER env var detected, forcing custom/external facts off to prevent recursion')
+        OptionStore.make_options_recursion_safe
+      end
       @internal_facts = load_internal_facts(user_query, options)
       @custom_facts = load_custom_facts(options)
       @external_facts = load_external_facts(options)
```

Comment:
Would it make sense to just not call `load_external_facts` if `INSIDE_FACTER` is set, like Facter 3 used to do: https://github.com/puppetlabs/facter/blob/6d7ffc6efdfbc3b1fc79311cdeb4581ac2098d9c/exe/facter.cc#L282-L292? Might be good to include that warning message too so people know how to fix the problem in their external facts.

That way you don't need to mutate the options.